### PR TITLE
[Workflow] improve workflow config validation

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by xabbuh** [comment]
We should add some tests too.

Status: Needs work

**Comment by xabbuh** [comment]
Tests added

Status: Needs review

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -258,13 +258,16 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
...
-                                    })
+                                    ->ifTrue(function ($v) { return !isset($v['type']) && !isset($v['service']); })
+                                    ->thenInvalid('One of "type" or "service" must be configured.')
+                                ->end()
+                                ->validate()
+                                    ->ifTrue(function ($v) { return isset($v['type']) && isset($v['service']); })
+                                    ->thenInvalid('"type" and "service" cannot be used together.')
+                                ->end()
+                                ->validate()
+                                    ->ifTrue(function ($v) { return isset($v['arguments']) && isset($v['service']); })
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: you should check whether ``arguments`` is an empty array or no actually, to account for people repeating the default value of the node explicitly (needed if they overwrite the definition in a subsequent file). An empty array should not trigger the exception.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    <framework:config>
+        <framework:workflows>
+            <framework:workflow name="my_workflow">
+                <framework:marking-store>
+                    <framework:type>multiple_state</framework:type>
+                    <framework:service>workflow_service</framework:service>
+                </framework:marking-store>
+                <framework:supports>Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\FrameworkExtensionTest</framework:supports>
+                <framework:places>first</framework:places>
+                <framework:places>last</framework:places>
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml]: the XML should be improved. This should be ``place``, not ``places``.

The Configuration class actually misses the necessary ``fixXmlConfig`` calls for workflows, places and transitions.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                <framework:transitions>
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml]: this node should not exist in the XML. It is not consistent with the way things work elsewhere.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    xmlns:framework="http://symfony.com/schema/dic/symfony"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
+                        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+    <framework:config>
+        <framework:workflows>
+            <framework:workflow name="my_workflow">
+                <framework:marking-store>
+                    <framework:type>multiple_state</framework:type>
+                    <framework:service>workflow_service</framework:service>
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml]: these should be attributes, not subnodes

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:framework="http://symfony.com/schema/dic/symfony"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
+                        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+    <framework:config>
+        <framework:workflows>
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml]: this node should not exist either

**Comment by stof** [comment]
If you are not comfortable with the XML config format, I can submit a separate PR instead to fix it.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml (hunk 2 above)
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflow_with_type_and_service.xml]: see #20458 which addresses most of the requested changes

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: An empty array is a valid value (and it *is* the default value of the node btw, as it is a prototyped array node)

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I cannot confirm that. `requiresAtLeastOneElement()` works as expected when playing with the fixtures I added in this PR.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -258,13 +259,16 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
...
                                 ->end()
                                 ->validate()
-                                    ->always(function ($v) {
-                                        if (isset($v['type']) && isset($v['service'])) {
-                                            throw new \InvalidArgumentException('"type" and "service" could not be used together.');
-                                        }
-
-                                        return $v;
-                                    })
+                                    ->ifTrue(function ($v) { return !isset($v['type']) && !isset($v['service']); })
```
Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I my previous PR, I added that too. But I realized it's not mandatory. If nothing is configured, we don't give a MarkingStore to the workflow, and so the workflow will use the default MarkingStore. So it makes the configuration simpler.

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: You are right. I reverted this change.

**Comment by lyrixx** [comment]
@xabbuh I have merged other PRs relative to the configuration. Could you rebase this one? Thanks ;)

**Comment by xabbuh** [comment]
@lyrixx Should I wait for #20458 to avoid rebasing twice or do you prefer to merge here first?

**Comment by lyrixx** [comment]
Indeed. I forget the other one. I prefer to wait another +1 there before the merge.

**Comment by stof** [comment]
I merged #20458 so please update your XML fixtures

**Comment by xabbuh** [comment]
I rebased and updated the XML fixtures.

**Comment by lyrixx** [comment]
:+1:

**Comment by lyrixx** [comment]
Thank you @xabbuh.

**Pull request merged**